src/tls_relation.py

```diff
@@ -0,0 +1,318 @@
+            logger.exception("Private key for hostname: %s not found", hostname)
+            raise GetPrivateKeyError(f"Private key for hostname: {hostname} not found") from exc
+        return KeyPair(private_key, password)
+
+    def get_provider_cert_with_hostname(
+        self, hostname: str
+    ) -> typing.Optional[ProviderCertificate]:
+        """Get a cert from the provider's integration data that matches 'certificate'.
+
+        Args:
+            hostname: the hostname to match with provider certificates
+
+        Returns:
+            typing.Optional[ProviderCertificate]: ProviderCertificate if exists, else None.
+        """
+        for cert in self.certificates.get_provider_certificates():
+            if get_hostname_from_cert(cert.certificate) == hostname:
+                return cert
+        return None
+
+    def certificate_expiring(self, certificate: str) -> None:
+        """Handle the TLS Certificate expiring event.
+
+        Generate a new CSR and request for a new certificate.
+
+        Args:
+            certificate: The invalidated certificate.
+        """
+        if expiring_cert := self._get_cert(certificate):
+            hostname = get_hostname_from_cert(expiring_cert.certificate)
+            old_csr = expiring_cert.csr
+            private_key, password = self._get_private_key(hostname)
+            new_csr = generate_csr(
+                private_key=private_key.encode(),
+                private_key_password=password.encode(),
+                subject=hostname,
+                sans_dns=[hostname],
+            )
+            self.certificates.request_certificate_renewal(
+                old_certificate_signing_request=old_csr.encode(),
+                new_certificate_signing_request=new_csr,
+            )
+
+    def certificate_invalidated(
+        self,
+        certificate: typing.Optional[str] = None,
+        provider_certificate: typing.Optional[ProviderCertificate] = None,
+    ) -> None:
+        """Handle TLS Certificate revocation.
+
+        Args:
+            certificate: The invalidated certificate to match with a provider certificate.
+            provider_certificate: The provider certificate, skip certificate matching
+            if this is provided directly.
+        """
+        if not certificate and not provider_certificate:
+            return
+
+        invalidated_provider_certificate = provider_certificate
+        if provider_certificate is None:
+            # certificate should always be set here
+            assert certificate  # nosec: B101
+
+            matched_provider_certificate = self._get_cert(certificate)
+            if not matched_provider_certificate:
+                logger.error("Cannot find provider certificate with provided certificate string.")
+                return
+            invalidated_provider_certificate = matched_provider_certificate
+
+        # For mypy only
+        # invalidated_provider_certificate should always be set here
```

Comment:
If it's just for me, couldn't we use a cast?

Reply:
good point, I changed this to a cast